### DISCUSSION TITLE
Changing storage during runtime

### DISCUSSION THREAD

**Discussion by Mobotium** [TokisanGames/Terrain3D#428] [Questions] [answered]
I'm attempting to change the map that a Terrain3D node is rendering during runtime.
Current attempt looks somewhat like this:

``
Resource res = ResourceLoader.Load("MapStorage.res");
``
``
terrain_3d.Set("storage", res);
``

Which leads to a crash whenever this is attempted.

Is there currently a way to alter the terrain display during runtime? The alternative would be creating scenes for each map and instancing them, which could be questionable depending on how many are needed.

**Comment by TokisanGames** [comment]
The editor is runtime and it can swap in and out different storage files without issue. There's no difference. The editor is certainly doing something you could do but are not. 

> Resource res = ResourceLoader.Load("MapStorage.res");
terrain_3d.Set("storage", res);

You can load that file and the other file in the editor, and swap them in and out without issue? I would type it Terrain3DStorage, and use set_storage().

> Which leads to a crash whenever this is attempted.

At the least, it shouldn't crash. You can enable debug logging and trace activity before the crash. You can also compare the log between changing res files in the editor vs your code to see what is different. If you're building from source you can debug and capture the exact line it breaks on. 
 
> The alternative would be creating scenes for each map and instancing them

We're doing this.

Storage is changing dramatically in PR #374 and 0.9.3.

**Reply by Mobotium** [reply]
> You can load that file and the other file in the editor, and swap them in and out without issue?

No issue at all, bar the weirdness with regions when the storage is loaded on multiple Terrain3d, which I saw an issue for already.

> I would type it Terrain3DStorage, and use set_storage().

I have not yet found a way for VSCode to recognize the classes from this addon, hence the weird way of doing it.
Regardless, I did attempt using set_storage instead and it resulted in the same outcome.

> You can enable debug logging and trace activity before the crash.

Have tried it by setting log to debug on the Terrain3D object in the editor.
However, the crash seems to be instant, and nothing is printed. Not even the initial "Set storage"

> Storage is changing dramatically in PR https://github.com/TokisanGames/Terrain3D/pull/374 and 0.9.3.

Good to know.
Might just wait for that, in case there's no obvious solution to this issue.

**Reply by TokisanGames** [reply]
Try setting it to null first and/or Terrain3DStorage.new() before loading the new one.
Also _when_ are you setting it? In ready? Is it attached to the tree?

**Reply by TokisanGames** [reply, marked as answer]
I added this to the bottom of `CodeGenerated.gd` and it worked just fine. There was also no issue if I commented the timer (processing on the same frame).
```
	await get_tree().create_timer(3).timeout		
	print("Loading new terrain")
	var res: Resource = ResourceLoader.load("res://demo/data/terrain_storage.res")
	terrain.set("storage", res)
```

Your issue might be with Godot's C# code and you probably need to do some debugging with both the Terrain3D and Godot source. When setup, it's pretty easy to run through the MSVC debugger and have it stop on the exact line with the issue. You want to confirm if we have a problem in Terrain3D. I think the above strongly suggests no.

**Reply by Mobotium** [reply]
> Also when are you setting it? In ready? Is it attached to the tree?

Have tried both, getting the same result.

Took a while to setup C# debugging, but got it working.
Problem seems to be some kind of unhandled exception in Godot's C# Callable:

> Exception has occurred: CLR/System.Runtime.InteropServices.SEHException
An unhandled exception of type 'System.Runtime.InteropServices.SEHException' occurred in GodotSharp.dll: 'External component has thrown an exception.'
<Cannot evaluate the exception stack trace>

Hard to get any other info out of it since the exception itself seems empty. Also can't say wherever "External component" could possibly refer to Terrain3D.

Regardless, does appear to be an issue with Godot C# with your working example.

I'll see if I can debug further.

**Reply by TokisanGames** [reply]
I'm able to debug Godot, Godot-cpp, and Terrain3D with source symbols via msvc. All three need to be built with debug symbols. I build and run the engine, with the debugger loading the project with the command line so the debugger is watching at startup, and it loads the symbols and code for all three projects. I can catch exceptions in any project without the external components problem.

**Reply by TokisanGames** [reply]
Closing due to this being Godot's issue, since runtime storage changing works fine in C++ or GDScript. We can reopen if need be.
